lib/intl.ts

```diff
@@ -45,36 +55,79 @@ const ObjectAssign = Object.assign;
+  [YM]: typeof yearMonthAmend | typeof globalThis.Intl.DateTimeFormat;
+  [MD]: typeof monthDayAmend | typeof globalThis.Intl.DateTimeFormat;
+  [TIME]: typeof timeAmend | typeof globalThis.Intl.DateTimeFormat;
+  [DATETIME]: typeof datetimeAmend | typeof globalThis.Intl.DateTimeFormat;
+  [ZONED]: typeof zonedDateTimeAmend | typeof globalThis.Intl.DateTimeFormat;
+  [INST]: typeof instantAmend | typeof globalThis.Intl.DateTimeFormat;
+}
+
+interface PrivateProps extends CustomFormatters {
+  [ORIGINAL]: globalThis.Intl.DateTimeFormat;
+  [TZ_RESOLVED]: Temporal.TimeZoneProtocol | string;
+  [TZ_GIVEN]: Temporal.TimeZoneProtocol | string | null;
+  [CAL_ID]: globalThis.Intl.ResolvedDateTimeFormatOptions['calendar'];
+  [LOCALE]: globalThis.Intl.ResolvedDateTimeFormatOptions['locale'];
+  [OPTIONS]: Intl.DateTimeFormatOptions;
+}
+
+type OptionsAmenderFunction = (options: Intl.DateTimeFormatOptions) => globalThis.Intl.DateTimeFormatOptions;
+type FormatterOrAmender = globalThis.Intl.DateTimeFormat | OptionsAmenderFunction;
+
 // Construction of built-in Intl.DateTimeFormat objects is sloooooow,
 // so we'll only create those instances when we need them.
 // See https://bugs.chromium.org/p/v8/issues/detail?id=6528
-function getPropLazy(obj, prop) {
-  let val = obj[prop];
+function getPropLazy<T extends PrivateProps, P extends keyof CustomFormatters>(
+  obj: T,
+  prop: P
+): globalThis.Intl.DateTimeFormat {
+  let val = obj[prop] as FormatterOrAmender;
   if (typeof val === 'function') {
+    // If we get here, `val` is an "amender function". It will take the user's
+    // options and transform them into suitable options to be passed into the
+    // built-in (non-polyfill) Intl.DateTimeFormat constructor. These options
+    // will vary depending on the Temporal type, so that's why we store separate
+    // formatters in separate props on the polyfill's DateTimeFormat instances.
+    // The efficiency happens because we don't create an (expensive) formatter
+    // until the user calls toLocaleString for that Temporal type.
     val = new IntlDateTimeFormat(obj[LOCALE], val(obj[OPTIONS]));
-    obj[prop] = val;
+    // TODO: can this be typed more cleanly?
+    (obj[prop] as globalThis.Intl.DateTimeFormat) = val;
   }
   return val;
 }
+
 // Similarly, lazy-init TimeZone instances.
-function getResolvedTimeZoneLazy(obj) {
+function getResolvedTimeZoneLazy(obj: PrivateProps) {
   let val = obj[TZ_RESOLVED];
   if (typeof val === 'string') {
-    val = new TimeZone(val);
+    val = ES.ToTemporalTimeZone(val);
     obj[TZ_RESOLVED] = val;
   }
   return val;
 }
 
-export function DateTimeFormat(
-  this: Intl.DateTimeFormat,
-  locale = undefined,
-  optionsParam: Partial<Intl.DateTimeFormatOptions> = {}
-): void {
-  if (!(this instanceof DateTimeFormat)) return new DateTimeFormat(locale, optionsParam);
+type DateTimeFormatImpl = Intl.DateTimeFormat & PrivateProps;
+
+function DateTimeFormatImpl(
```

Comment:
Here's the story behind renaming this function: I wanted to only export this function as an Intl.DateTimeFormat (from the public types) so that other modules wouldn't know about its implementation details. This is what we do, for example, with other Temporal types where callers can `import { Temporal } from '..'`.  (BTW, this works in tests but we should validate that it will also work in real apps!)

But the module resolver used in Demitasse tests (resolve.source.mjs) was special-cased for importing Temporal only. `import { Temporal, Intl } from '..'` caused Demitasse tests to fail. 

But I still wanted other-module callers to see this function as a plain `Intl.DateTimeFormat` (from public types) without seeing its private properties. I couldn't figure out how to do this when it was named `DateTimeFormat` without getting a TS error when attempting to change its type to the public type. So I renamed the function, which required a runtime code change to update the `name` property.

Feel free to suggest a better way to do it!